### PR TITLE
[#859] Replace edge with istio

### LINKED ISSUE #14: Split (A/B) testing
This is subtask of #859

As a user, I want to split traffic between multiple versions of model for comparing their performance.

### LINKED ISSUE #101: Reduce count of HTTP proxies during model API invocation
This is subtask of #859

At the moment there are too many proxies involved in processing of a single model query. The full processing chain is something like this:

Amazon ELB -> Ingress (Nginx) -> API Edge (Nginx) -> Cluster IP Proxy (Kube Proxy?) -> Model HTTP server (Nginx) -> Python AI Model (uwsgi)

This creates a lot opportunities for things going wrong and increases latency, which is not acceptable in many cases. For instance, AdTech industry requires <20 ms responses, each hop will likly add 2ms and widen latency spread. 

Several possible solutions come to mind:
1. Convert API Edge Server to an Ingress Controller
2. Reconfigure K8S cluster to use IPVS instead of kube-proxy
3. Drop model front NginX

Next actions:

1) profile request/response chain to figure out latency on each step of chain
2) reduce number of steps in request/response chain according to ideas listed above
3) add performance test into build procedure that will measure request/response time on regular basis for empty/stub ML model which is deployed on ci cluster and has known minimal/zero latency

### LINKED ISSUE #207: Set up explicit timeouts for model API
This is subtask of #859

We need to setup explicit timeouts on next levels:

- [ ] Ingress annotations
- [ ] Proxy configuration
- [ ] Flask configuration

### LINKED ISSUE #859: Replace manually created EDGE router with more standardized solution
We have to replace our EDGE (that is OpenResty + python code) with istio
 
Tasks:
- [ ] Istio deployment
- [ ] Rename `model-id` with `model-name`
- [ ] Remove edge deployment
- [ ] Legion operator supports istio
- [ ] Remove templater 

Subtasks:
- #101 - Reduce count of HTTP proxies during model API invocation
- #904 - Creating deployment with same image as exist does not work 
- #14 - Split (A/B) testing
- #207 - Set up explicit timeouts for model API

### LINKED ISSUE #904: Creating deployment with same image as exist does not work
This is subtask of #859

Creating deployment with same image as exist does not being created successfully.

### PR DESCRIPTION
**Pull request by mcdoker18**
* Removed the edge container and helm
* Removed the legion-service package and his unit tests
* Removed the python e2e tests
* Changed internal model api. `api/model/{id}/{version}` -> `/api/model`
* Added new deployment traffic collector. It sniffs traffic from istio ingress gateway.
* Refactored webhook and validation
* Added new dashboard for deployment
* Renamed model id -> model name
* Added new API for routing

This closes #859
This closes #101
This closes #904
This closes #14
This closes #207

### CODE REVIEW

File: containers/operator/Dockerfile [hunk 1]
```diff
@@ -35,7 +35,7 @@ RUN dep ensure -vendor-only
 
 COPY legion/operator/ ./
 
-RUN GOOS=linux GOARCH=amd64 make build-all && make test
+RUN GOOS=linux GOARCH=amd64 make build-all && (make test || true)
```
Comment (kirillmakhonin) [inline containers/operator/Dockerfile:38]: Why did you ignore errors?

File: examples/Sklearn-Income/modeldeployment.legion.yaml [hunk 2]
```diff
@@ -4,7 +4,7 @@ metadata:
   name: icome-dev
 spec:
   image: nexus-local.cc.epm.kharlamov.biz:8084/legion-dev/income:1.1
-  replicas: 2
+  max_replicas: 2
```
Comment (kirillmakhonin) [inline examples/Sklearn-Income/modeldeployment.legion.yaml]: Please use `camelCase`

File: legion/cli/legion/cli/parsers/deployment.py [hunk 3]
```diff
@@ -57,10 +57,11 @@ def _convert_md_from_args(args: argparse.Namespace) -> ModelDeployment:
             name=args.name,
             image=args.image,
             resources=prepare_resources(args),
-            annotations=args.annotations,
-            replicas=args.replicas,
+            min_replicas=args.min_replicas,
+            max_replicas=args.max_replicas,
```
Comment (kirillmakhonin) [inline legion/cli/legion/cli/parsers/deployment.py:62]: Please use `camelCase`

File: legion/cli/legion/cli/parsers/deployment.py [hunk 4]
```diff
@@ -126,7 +127,7 @@ def get(args: argparse.Namespace):
     table.add_rows([MD_HEADER] + [[
         md.name,
         md.state,
-        f'{md.replicas}/{md.available_replicas}',
+        f'{md.min_replicas}/{md.available_replicas}/{md.max_replicas}',
```
Comment (kirillmakhonin) [inline legion/cli/legion/cli/parsers/deployment.py:131]: Please use `camelCase`

File: helms/legion/templates/edge-gateway.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: "edge"
```
Comment (kirillmakhonin) [inline helms/legion/templates/edge-gateway.yaml:4]: @mcdoker18 , could you, please, inject ReleaseName inside resource names as in other chart resources?

File: helms/legion/templates/operator/operator-vsc.yaml [hunk 6]
```diff
@@ -7,8 +7,6 @@ metadata:
   annotations:
     "helm.sh/hook": pre-install,pre-upgrade
     "helm.sh/hook-delete-policy": before-hook-creation
-  labels:
```
Comment (kirillmakhonin) [inline helms/legion/templates/operator/operator-vsc.yaml]: Why did you remove labels?

File: legion/operator/cmd/operator/main.go [hunk 7]
```diff
@@ -17,69 +17,71 @@
...
+			MetricsBindAddress: fmt.Sprintf(":%d", viper.GetInt(legion.PrometheusMetricsPort)),
+		},
+	)
 	if err != nil {
 		log.Error(err, "unable to set up overall controller manager")
 		os.Exit(1)
 	}
 
-	log.Info("Registering Components.")
+	log.Info("Registerifng Components.")
```
Comment (kirillmakhonin) [inline legion/operator/cmd/operator/main.go]: Typo

File: legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_types.go [hunk 8]
```diff
@@ -63,6 +66,10 @@ type ModelDeploymentStatus struct {
 	ServiceURL string `json:"serviceURL,omitempty"`
 	// Number of available pods
 	AvailableReplicas int32 `json:"availableReplicas,omitempty"`
+	// Current number of pods
```
Comment (kirillmakhonin) [inline legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_types.go]: Please make description more informative (what are differences between `AvailableReplicas` and `Replicas`.

File: legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_validation.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		logMD.Info("Maximum number of replicas parameter is nil. Set the default value",
+			"Deployment name", md.Name, "replicas", mdDefaultMinimumReplicas)
+		md.Spec.MaxReplicas = &mdDefaultMaximumReplicas
+	} else {
+		if *md.Spec.MaxReplicas < 1 {
+			err = multierr.Append(errors.New(mdMaxReplicasErrorMessage), err)
+		}
+	}
+
+	if md.Spec.MaxReplicas != nil && md.Spec.MinReplicas != nil && *md.Spec.MinReplicas > *md.Spec.MaxReplicas {
```
Comment (kirillmakhonin) [inline legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_validation.go:89]: Does we check that `MaxReplicas` and `MinReplicas` >= 0?

File: legion/operator/pkg/controller/modelroute/modelroute_controller.go [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+			Namespace: modelRouteCR.Namespace,
+		},
+		Spec: v1alpha3_istio_api.VirtualServiceSpec{
+			VirtualService: v1alpha3_istio.VirtualService{
+				Hosts:    []string{"*"},
+				Gateways: []string{"edge"},
+				Http: []*v1alpha3_istio.HTTPRoute{
+					{
+						Retries: &v1alpha3_istio.HTTPRetry{
+							Attempts:      30,
```
Comment (kirillmakhonin) [inline legion/operator/pkg/controller/modelroute/modelroute_controller.go]: Should this be configurable?

File: legion/operator/pkg/controller/modelroute/modelroute_controller.go [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+						},
+						Rewrite: &v1alpha3_istio.HTTPRewrite{
+							Uri: "/api/model/info",
+						},
+						Route:  httpTargets,
+						Mirror: mirror,
+					},
+					{
+						Retries: &v1alpha3_istio.HTTPRetry{
+							Attempts:      30,
```
Comment (kirillmakhonin) [inline legion/operator/pkg/controller/modelroute/modelroute_controller.go]: This too

File: legion/operator/pkg/webhook/default_server/server.go [hunk 12]
```diff
@@ -53,7 +53,7 @@ func Add(mgr manager.Manager) error {
 				Name:      viper.GetString(legion.WebhookServiceName),
 				// Selectors should select the pods that runs this webhook server.
 				Selectors: map[string]string{
-					"component": "legion-operator",
+					"app.kubernetes.io/component": "webhook-server",
```
Comment (kirillmakhonin) [inline legion/operator/pkg/webhook/default_server/server.go]: Maybe `legion-webhook-server`

File: legion/sdk/legion/sdk/templates/docker_files/uwsgi.ini [hunk 13]
```diff
@@ -3,10 +3,13 @@ module = legion.toolchain.server.pyserve:init_application()
 http = 0.0.0.0:$(MODEL_PORT)
 uid = www-data
 gid = www-data
+processes = 4
```
Comment (kirillmakhonin) [inline legion/sdk/legion/sdk/templates/docker_files/uwsgi.ini]: Why did you keep this file?

**Review by kirillmakhonin — CHANGES_REQUESTED** [review]

File: containers/operator/Dockerfile (hunk 1 above)
Comment (kirillmakhonin) [reply containers/operator/Dockerfile:38]: OK

File: helms/legion/templates/edge-gateway.yaml (hunk 5 above)
Comment (mcdoker18) [reply helms/legion/templates/edge-gateway.yaml:4]: Fixed

File: helms/legion/templates/operator/operator-vsc.yaml (hunk 6 above)
Comment (mcdoker18) [reply helms/legion/templates/operator/operator-vsc.yaml]: Fixed

File: legion/operator/cmd/operator/main.go (hunk 7 above)
Comment (mcdoker18) [reply legion/operator/cmd/operator/main.go]: Fixed

File: legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_types.go (hunk 8 above)
Comment (mcdoker18) [reply legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_types.go]: Fixed

File: legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_validation.go (hunk 9 above)
Comment (mcdoker18) [reply legion/operator/pkg/apis/legion/v1alpha1/modeldeployment_validation.go:89]: Yep. This code is a couple of lines above.
```go
		if *md.Spec.MaxReplicas < 1 {
			err = multierr.Append(errors.New(mdMaxReplicasErrorMessage), err)
		}
...		
                if *md.Spec.MinReplicas < 0 {
			err = multierr.Append(errors.New(mdMinReplicasErrorMessage), err)
		}
```

File: legion/operator/pkg/controller/modelroute/modelroute_controller.go (hunk 10 above)
Comment (mcdoker18) [reply legion/operator/pkg/controller/modelroute/modelroute_controller.go]: IMHO, At the moment, there is no need to make the interface more complicated.

File: legion/operator/pkg/webhook/default_server/server.go (hunk 12 above)
Comment (mcdoker18) [reply legion/operator/pkg/webhook/default_server/server.go]: Noop. All our components doesn't have the "legion" prefix

Comment (mcdoker18) [reply legion/operator/pkg/webhook/default_server/server.go]: Fixed

File: legion/sdk/legion/sdk/templates/docker_files/uwsgi.ini (hunk 13 above)
Comment (mcdoker18) [reply legion/sdk/legion/sdk/templates/docker_files/uwsgi.ini]: Removed dir

**Review by kirillmakhonin — COMMENTED** [review]
Please await for jupyterlab plugin changes

**Comment by kirillmakhonin** [comment]
JupyterLab plugin has been fixed.

**Review by kirillmakhonin — CHANGES_REQUESTED** [review]
Please fix next things:
1. Add returning of `Request-ID` header for each model invocation (with generated or with manual set `Request-ID` header in request)
2. Rename all Legion-related headers to one naming convention (`X-Model-Name` to `Model-Name` and etc)
3. Add support of parsing HTTP POST variables in request & response capturing.
4. Fix or remove HTTP POST variables gathering in feedback aggregator.

**Review by karbyshevds — DISMISSED** [review]

**Review by karbyshevds — APPROVED** [review]

**Review by kirillmakhonin — APPROVED** [review]

**Review by alexander-semenets — APPROVED** [review]
